### ISSUE TITLE
Json writer adds another "}" at the end

### ISSUE THREAD

**Issue by PinkieSwirl** [ktorio/ktor#963]
### Ktor Version
1.1.2, 1.1.3
### Ktor Engine Used(client or server and name)
netty server
### JVM Version, Operating System and Relevant Context
JRE 11, Windows 10 / Ubuntu 18.04

```
sourceCompatibility = JavaVersion.VERSION_11
compileKotlin {
    kotlinOptions {
        jvmTarget = "1.8"
        freeCompilerArgs = [
                "-XXLanguage:+InlineClasses",
                "-Xuse-experimental=kotlin.ExperimentalUnsignedTypes",
                "-progressive",
                "-XXLanguage:+NewInference"
        ]
    }
}
```

### Feedback
When using the directory based session storage (with or without cache) an extra `}` is sometimes appended to the stringified session.

This does not happen with the in memory session!

Sometimes, since when the session is created (was not in the store) everything works just fine, but when I want to access the session later I get the error (shortened the json string):
```
java.lang.IllegalStateException: Reader has not consumed the whole input: JsonReader(source='{...}}', currentPosition=496, tokenClass=7, tokenPosition=495, offset=493)
	at kotlinx.serialization.json.Json.parse(Json.kt:122)
	at kotlinx.serialization.json.Json$Companion.parse(Json.kt:161)
	at de.rub.mobsec.changeling.JsonSessionSerializer.deserialize(Serialization.kt:15)
	at io.ktor.sessions.SessionTrackerById$load$2.invokeSuspend(SessionTrackerById.kt:32)
	at io.ktor.sessions.SessionTrackerById$load$2.invoke(SessionTrackerById.kt)
	at io.ktor.sessions.CacheStorage.read(CacheStorage.kt:17)
	at io.ktor.sessions.SessionTrackerById.load(SessionTrackerById.kt:30)
	at io.ktor.sessions.Sessions$Feature$install$1.invokeSuspend(Sessions.kt:45)
	at io.ktor.sessions.Sessions$Feature$install$1.invoke(Sessions.kt)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:278)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:137)
	at io.ktor.util.pipeline.SuspendFunctionGun.execute(PipelineContext.kt:157)
	at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:23)
	at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$2.invokeSuspend(DefaultEnginePipeline.kt:106)
	at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$2.invoke(DefaultEnginePipeline.kt)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:278)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:137)
	at io.ktor.util.pipeline.SuspendFunctionGun.execute(PipelineContext.kt:157)
	at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:23)
	at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1.invokeSuspend(NettyApplicationCallHandler.kt:31)
	at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1.invoke(NettyApplicationCallHandler.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startCoroutineUndispatched(Undispatched.kt:55)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:111)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:160)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:54)
	at kotlinx.coroutines.BuildersKt.launch(Unknown Source)
	at io.ktor.server.netty.NettyApplicationCallHandler.handleRequest(NettyApplicationCallHandler.kt:22)
	at io.ktor.server.netty.NettyApplicationCallHandler.channelRead(NettyApplicationCallHandler.kt:16)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:353)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
Relevant part from my Serialization.kt:
```
class JsonSessionSerializer : SessionSerializer {
    override fun serialize(session: Any): String = Json.stringify(ChangelingSession.serializer(),session as ChangelingSession).also { println(it) }

    override fun deserialize(text: String): Any = Json.parse(ChangelingSession.serializer(), text) // line 15
}
```

The `.also { println(it) }` statement never prints a string with two `}` at the end.

My current workaround:
```
    override fun deserialize(text: String): Any {
        return if(text[text.lastIndex] == '}' && text[text.lastIndex-1] == '}')
            Json.parse(ChangelingSession.serializer(), text.substring(0, text.lastIndex))
        else Json.parse(ChangelingSession.serializer(), text)
    }
```
This does work in my case, but has obvious flaws.

**Comment by PinkieSwirl** [comment]
My guess is that it is an concurrency issue. 

The problem got worse after I added more stuff to my session (=longer json string) and doubled the length of the session ID: Now it "appends" more than one character, it seems to be the last X characters, my current "best" is X = 2.

**Comment by cy6erGn0m** [comment]
No, it's not related to concurrency. Fix is here: https://github.com/ktorio/ktor/pull/968

**Comment by cy6erGn0m** [comment]
Fixed in 1.2.0-rc2
